crates/api/src/builder/api.rs

```diff
@@ -689,16 +671,42 @@ where
             )
             .await?;
 
-        // Fetch constraints
-        let constraints = api.auctioneer.get_constraints(payload.slot()).await?;
-        let Some(constraints) = constraints else {
-            warn!(request_id = %request_id, "no constraints found for slot");
-            return Err(BuilderApiError::NoConstraintsFound);
-        };
+        // Fetch constraints, and if available verify inclusion proofs and save them to cache
+        if let Some(constraints) = api.auctioneer.get_constraints(payload.slot()).await? {
+            let transactions_root: B256 = payload
+                .transactions()
+                .clone()
+                .hash_tree_root()?
+                .to_vec()
+                .as_slice()
+                .try_into()
+                .map_err(|e| {
+                    error!(error = %e, "failed to convert root to hash32");
+                    BuilderApiError::InternalError
+                })?;
+            let proofs = payload.proofs().ok_or(BuilderApiError::InclusionProofsNotFound)?;
+            let constraints_proofs: Vec<_> = constraints.iter().map(|c| &c.proof_data).collect();
+
+            verify_multiproofs(constraints_proofs.as_slice(), proofs, transactions_root).map_err(
+                |e| {
+                    error!(error = %e, "failed to verify inclusion proofs");
+                    BuilderApiError::InclusionProofVerificationFailed(e)
+                },
+            )?;
 
-        // Verify inclusion proofs
-        if let Err(err) = api.verify_inclusion_proof(&payload, &constraints).await {
-            return Err(err);
+            // Save inclusion proof to auctioneer.
+            api.save_inclusion_proof(
```

Comment:
can you briefly explain why you moved `save_inclusion_proof` here as opposed to line 748?

Reply:
Sure. In short we want to save proofs to cache only when:
- we have constraints
- we have proofs
- proofs are correct

Before we were already assuming proofs were available via `expect`, which may not be the case.

Reply:
Perfect